[Merged by Bors] - feat (NumberTheory/LSeries): Hurwitz zeta

loefflerd opened the pull request:
Add the Hurwitz zeta function (defined as the sum of its even and odd parts, which were already defined in previous PR's).

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+* `expZeta`: the exponential zeta function
+* `hasSum_nat_hurwitzZeta_of_mem_Icc` and `hasSum_expZeta_of_one_lt_re`:
+  relation to Dirichlet series for `1 < re s`
+* `differentiableAt_hurwitzZeta` and `differentiableAt_expZeta`: analyticity away from `s = 1`
```
```suggestion
* ` hurwitzZeta_residue_one` shows that the residue at `s = 1` equals `1`
* `differentiableAt_hurwitzZeta` and `differentiableAt_expZeta`: analyticity away from `s = 1`
```

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+
+lemma hurwitzZetaEven_eq (a : UnitAddCircle) (s : ℂ) :
+    hurwitzZetaEven a s = (hurwitzZeta a s + hurwitzZeta (-a) s) / 2 := by
+  simp_rw [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```
```suggestion
  simp only [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+
+lemma hurwitzZetaOdd_eq (a : UnitAddCircle) (s : ℂ) :
+    hurwitzZetaOdd a s = (hurwitzZeta a s - hurwitzZeta (-a) s) / 2 := by
+  simp_rw [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```
```suggestion
  simp only [hurwitzZeta, hurwitzZetaEven_neg, hurwitzZetaOdd_neg]
```

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+/-- The residue of the Hurwitz zeta function at `s = 1` is `1`. -/
+lemma hurwitzZeta_residue_one (a : UnitAddCircle) :
+    Tendsto (fun s ↦ (s - 1) * hurwitzZeta a s) (𝓝[≠] 1) (𝓝 1) := by
+  simp_rw [hurwitzZeta, mul_add, (by simp : 𝓝 (1 : ℂ) = 𝓝 (1 + (1 - 1) * hurwitzZetaOdd a 1))]
```
```suggestion
  simp only [hurwitzZeta, mul_add, (by simp : 𝓝 (1 : ℂ) = 𝓝 (1 + (1 - 1) * hurwitzZetaOdd a 1))]
```

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+determining what that value is). -/
+lemma tendsto_hurwitzZeta_sub_one_div_nhds_one (a : UnitAddCircle) :
+    Tendsto (fun s ↦ hurwitzZeta a s - 1 / (s - 1) / Gammaℝ s) (𝓝 1) (𝓝 (hurwitzZeta a 1)) := by
+  simp_rw [hurwitzZeta, add_sub_right_comm]
```
```suggestion
  simp only [hurwitzZeta, add_sub_right_comm]
```

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+/-- The difference of two Hurwitz zeta functions is differentiable everywhere. -/
+lemma differentiable_hurwitzZeta_sub_hurwitzZeta (a b : UnitAddCircle) :
+    Differentiable ℂ (fun s ↦ hurwitzZeta a s - hurwitzZeta b s) := by
+  simp_rw [hurwitzZeta, add_sub_add_comm]
```
```suggestion
  simp only [hurwitzZeta, add_sub_add_comm]
```

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+## The functional equation
+-/
+
+/-- Functional equation for Hurwitz zeta function. -/
```
```suggestion
/-- Functional equation for the Hurwitz zeta function. -/
```

faenuccio commented on Mathlib/NumberTheory/LSeries/HurwitzZeta.lean:
```diff
@@ -0,0 +1,164 @@
+    expZeta, expZeta, Complex.cos, Complex.sin, sinZeta_neg, cosZeta_neg]
+  ring_nf
+
+/-- Functional equation for exponential zeta function. -/
```
```suggestion
/-- Functional equation for the exponential zeta function. -/
```

faenuccio commented:
LGTM!

loefflerd commented:
Thanks for the review, Filippo! I have applied your changes. I also fixed a minor formatting glitch in the module docstring, and added a mssing lemma `LSeriesHasSum_exp` that makes the link with Michael's `LSeries` API.

faenuccio commented:
> Thanks for the review, Filippo! I have applied your changes. I also fixed a minor formatting glitch in the module docstring, and added a mssing lemma `LSeriesHasSum_exp` that makes the link with Michael's `LSeries` API.

Great!

loefflerd commented:
Hi Michael,

Thanks for the suggestions! I'm in two minds about this sort of thing (trading off concision / elegance vs. speed). CPU cycles are cheap, after all; and there is the very good point raised in the Zulip discussion "To automate or not not automate?", that making explicit workarounds for slow tactics makes it less likely that tactic-development people will notice and work on the slowness. 

Hence I've put in your `generalize` changes, but with a comment pointing out that it would be nice if `ring_nf` could do these things automatically. Are you happy with that compromise?

Regards, David

MichaelStollBayreuth commented:
I do understand that (and my suggestions were meant as such, not as requirements).
I think the compromise (adding comments) is reasonable.
Of course, a better solution would be to make `ring_nf` faster (along similar lines, if at all  possible), but this is beyond my paygrade...

MichaelStollBayreuth reviewed (approved):
LGTM!

MichaelStollBayreuth commented:
maintainer merge

jcommelin commented:
Thanks :tada:

bors merge
